psi4/src/psi4/libmints/thc_eri.cc

```diff
@@ -0,0 +1,374 @@
+        use_df_ = false;
+    } else if (options_.get_bool("LS_THC_DF") && !auxiliary_) {
+        throw PSIEXCEPTION("    Attempting to do LS-THC with DF integrals w/o specifed auxiliary basis set :(\n\n");
+    }
+
+    print_header();
+
+    size_t nbf = primary_->nbf();
+
+    timer_on("LS-THC: Build Grid");
+
+    // Create a grid for LS_THC
+    std::map<std::string, std::string> grid_init_str_options = {
+        {"DFT_PRUNING_SCHEME", options_.get_str("LS_THC_PRUNING_SCHEME")},
+        {"DFT_RADIAL_SCHEME",  "TREUTLER"},
+        {"DFT_NUCLEAR_SCHEME", "TREUTLER"},
+        {"DFT_GRID_NAME",      ""},
+        {"DFT_BLOCK_SCHEME",   "OCTREE"},
+    };
+    std::map<std::string, int> grid_init_int_options = {
+        {"DFT_SPHERICAL_POINTS", options_.get_int("LS_THC_SPHERICAL_POINTS")}, 
+        {"DFT_RADIAL_POINTS",    options_.get_int("LS_THC_RADIAL_POINTS")},
+        {"DFT_BLOCK_MIN_POINTS", 100},
+        {"DFT_BLOCK_MAX_POINTS", 256},
+    };
+    std::map<std::string, double> grid_init_float_options = {
+        {"DFT_BASIS_TOLERANCE",   options_.get_double("LS_THC_BASIS_TOLERANCE")}, 
+        {"DFT_BS_RADIUS_ALPHA",   1.0},
+        {"DFT_PRUNING_ALPHA",     1.0},
+        {"DFT_BLOCK_MAX_RADIUS",  3.0},
+        {"DFT_WEIGHTS_TOLERANCE", options_.get_double("LS_THC_WEIGHTS_TOLERANCE")},
+    };
+    auto grid = DFTGrid(molecule_, primary_, grid_init_int_options, grid_init_str_options, grid_init_float_options, options_);
+    size_t npoints = grid.npoints();
+
+    x1_ = std::make_shared<Matrix>(npoints, nbf);
+
+    // Parrish 2012, Equation 29
+    size_t point_idx = 0;
+    for (auto& block : grid.blocks()) {
+        auto w = block->w();
+        auto x = block->x();
+        auto y = block->y();
+        auto z = block->z();
+
+#pragma omp parallel for
+        for (size_t p = 0; p < block->npoints(); ++p) {
+            primary_->compute_phi(&(*x1_)(point_idx + p, 0), x[p], y[p], z[p]);
+            x1_->scale_row(0, point_idx + p, std::pow(std::abs(w[p]), 0.25));
+        }
+
+        point_idx += block->npoints();
+    }
+
+    x2_ = x1_;
+    x3_ = x1_;
+    x4_ = x1_;
+
+    timer_off("LS-THC: Build Grid");
+    
+    timer_on("LS-THC: Form E");
+
+    // Parrish 2012, Equations 33-35
+    SharedMatrix E_PQ = use_df_ ? build_E_df() : build_E_exact();
+
+    timer_off("LS-THC: Form E");
+
+    timer_on("LS-THC: Form S");
+
+    SharedMatrix S_Qq = std::make_shared<Matrix>(npoints, npoints);
+    SharedMatrix S_temp = linalg::doublet(x1_, x1_, false, true);
+
+    auto S_Qq_p = S_Qq->pointer();
+    auto Stp = S_temp->pointer();
+
+    // Parrish 2012, Equation 28
+#pragma omp parallel for
+    for (size_t p = 0; p < npoints; ++p) {
+        for (size_t q = 0; q < npoints; ++q) {
+            S_Qq_p[p][q] = Stp[p][q] * Stp[p][q];
+        }
+    }
+
+    // Parrish 2012, Equation 30
+    int nremoved = 0;
+    SharedMatrix S_Qq_inv = S_Qq->pseudoinverse(options_.get_double("LS_THC_S_EPSILON"), nremoved);
```

Comment:
FYI: I've run 10 water molecules 240 AO basis, based on `test_thc_eri.py`, and it seems

`Matrix->pseudoinverse` bloats memory, more than I would expected.

mprof (rough idea of process memory):

![thc_01](https://github.com/psi4/psi4/assets/32617178/4212e085-d17c-4c39-9c64-14a0eb7fdc3d)

memray HTML (.tar.gz)